tests/languages/download_test.py

```diff
@@ -0,0 +1,191 @@
+from __future__ import annotations
+
+import _thread as thread
+import platform
+import stat
+from http.server import BaseHTTPRequestHandler
+from http.server import HTTPServer
+from os import chmod
+from pathlib import Path
+from typing import Iterator
+
+import pytest
+
+import pre_commit.constants as C
+from pre_commit import lang_base
+from pre_commit.languages import download
+from pre_commit.languages.download import ChecksumMismatchError
+from pre_commit.languages.download import ENVIRONMENT_DIR
+from pre_commit.languages.download import get_default_version
+from pre_commit.prefix import Prefix
+from testing.language_helpers import run_language
+
+shellscript = b'#!/bin/sh\necho hello\nexit 123'
+shell_checksum = 'oRJkj6Cr8nWIivZ9d3W+rVZt/aSW1l9YtxSVh+GtIHM='
+batch_script = b'@echo off\necho hello\nexit 123'
+batch_checksum = 'L63Nefq+fKVIm24IKqlcqbJmc1rrJD3dKhIvutFK+IA='
+if platform.system() == 'Windows':
+    correct_checksum = batch_checksum
+else:
+    correct_checksum = shell_checksum
+
+healthy_dependencies = [
+    f"""linux/amd64
+sha256-{shell_checksum}
+http://127.0.0.1:5555
+test.bat""",
+    f"""windows/amd64
+sha256-{batch_checksum}
+http://127.0.0.1:5555
+test.bat""",
+]
+
+deps_invalid_length = [
+    """linux/amd64
+sha256-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+http://127.0.0.1:5555
+test.bat""",
+    """windows/amd64
+sha256-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+http://127.0.0.1:5555
+test.bat""",
+]
+
+deps_wrong_checksum = [
+    """linux/amd64
+sha256-oRJkj6Cr8nWIivZ9d4W+rVZt/aSW1l9YtxSVh+GtIHM=
+http://127.0.0.1:5555
+test.bat""",
+    """windows/amd64
+sha256-L63Nefq+fKVIm25IKqlcqbJmc1rrJD3dKhIvutFK+IA=
+http://127.0.0.1:5555
+test.bat""",
+]
+
+
+class HTTPRequestHandler(BaseHTTPRequestHandler):
+    def do_GET(self):
+        self.send_response(200)
+        self.end_headers()
+        if platform.system() == 'Windows':
+            self.wfile.write(batch_script)
+        else:
+            self.wfile.write(shellscript)
+
+
+class server():
```

Comment:
* Use the `threading` rather than the low-level `_thread` module.
* Use a context manager fixture so that the server _always_ gets cleaned up

```py
@fixture
def server() -> Iterator[Uri]:
    httpd = HTTPServer(("localhost", 5555), HTTPRequestHandler)
    thread = Thread(target=httpd.serve_forever, name="HTTP Server")
    thread.start()
    host, port = httpd.server_address
    try:
        yield Uri(f"http://{host}:{port}")
    finally:
        httpd.shutdown()
        httpd.server_close()
        thread.join(1)
        assert not thread.is_alive()
```

You can even combine fixtures to auto-generate the additional dependencies, which makes sure the URL matches up. Usually in testing, fixtures are preferred over global variables.

```suggestion
@dataclass(frozen=True)
class Script:
    content: bytes
    integrity: Integrity

@fixture
def shell() -> Script:
    content= b'#!/bin/sh\necho hello\nexit 123'
    integrity = Integrity('sha256-oRJkj6Cr8nWIivZ9d3W+rVZt/aSW1l9YtxSVh+GtIHM=')
    return Script(content, integrity)

@fixture
def batch() -> Script:
    content= b'@echo off\necho hello\nexit 123'
    integrity = Integrity(''sha256-L63Nefq+fKVIm24IKqlcqbJmc1rrJD3dKhIvutFK+IA='')
    return Script(content, integrity)

@fixture
def script(shell: Script, batch: Script) -> Script:
    if platform.system() == 'Windows':
        return batch
    else:
        return shell

@dataclass(frozen=True)
class Server:
    uri: Uri
    script: Script

@fixture
def server(script: Script) -> Iterator[Server]:
    class HTTPRequestHandler(BaseHTTPRequestHandler):
        def do_GET(self):
            self.send_response(200)
            self.end_headers()
            self.wfile.write(script.content)

    httpd = HTTPServer(("localhost", 5555), HTTPRequestHandler)
    thread = Thread(target=httpd.serve_forever, name="HTTP Server")
    thread.start()
    host, port = httpd.server_address
    uri = Uri(f"http://{host}:{port}")
    try:
        yield Server(uri, fixture)
    finally:
        httpd.shutdown()
        httpd.server_close()
        thread.join(1)
        assert not thread.is_alive()

@dataclass(frozen=True)
class Fixture:
    server: Server
    dependencies: Collection[str]

@fixture
def health(server: Server, shell: Fixture, batch: Fixture) -> Fixture:
    dependencies = (
        f"""linux/amd64
{shell.integrity}
http://{server}
test""",
        f"""windows/amd64
{batch.integrity}
http://{server}
test""",
    )
    return Fixture(server, deps)

@fixture
def prefix(tmpdir) -> Prefix:
    with tmpdir.as_cwd():
        directory = tmpdir.join('prefix').ensure_dir()
        prefix = Prefix(str(directory))
        yield prefix
```

You can then use this like so:

```py
def test_healthy(prefix: Prefix, healthy: Fixture) -> None:
    # Do the test, the server will always be cleaned up.
    download.install_environment(prefix, C.DEFAULT, healthy.dependencies)
    assert download.health_check(prefix, C.DEFAULT) is None
```

The complexity is then contained in fixtures and the tests become extremely readable and safe.

Reply:
> Use the threading rather than the low-level _thread module.

Done with that point.